README.md
# docker-junit-rule [ ![Download](https://api.bintray.com/packages/klousiaj/maven/docker-junit-rule/images/download.svg) ](https://bintray.com/klousiaj/maven/docker-junit-rule/_latestVersion) [![Build Status](https://travis-ci.org/klousiaj/docker-junit-rule.svg)](https://travis-ci.org/klousiaj/docker-junit-rule) [![codecov.io](https://codecov.io/gh/klousiaj/docker-junit-rule/coverage.svg?branch=master)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=master)

A junit rule to run docker containers. This repository is based on, and extended from the excellent 
library written by Geoffroy Warin. You can find the original code [here](https://github.com/geowarin/docker-junit-rule).

## Versions
Check out the [ChangeLog](./CHANGELOG.md)

## Installation

This is hosted on bintray.

### Maven

Add the following to your `pom.xml`:

```xml
<repositories>
    <repository>
        <snapshots>
            <enabled>false</enabled>
        </snapshots>
        <id>central</id>
        <name>bintray</name>
        <url>http://jcenter.bintray.com</url>
    </repository>
</repositories>

...

<dependency>
    <groupId>com.github.klousiaj</groupId>
    <artifactId>docker-junit-rule</artifactId>
    <version>1.3.5</version>
    <scope>test</scope>
</dependency>
```

### Gradle

Add the following to your `build.gradle`:

```groovy
repositories {
  jcenter()
}

dependencies {
  testCompile 'com.github.klousiaj:docker-junit-rule:1.3.5'
}
```

## Usage

Example for rabbitMQ:

```java
import DockerRule;
import com.rabbitmq.client.ConnectionFactory;
import org.junit.ClassRule;
import org.junit.Test;

public class RabbitIntegrationTest {

  @ClassRule
  public static DockerRule rabbitRule =
    DockerRule.builder()
      .image("rabbitmq:management")
      .ports("5672", ":32779", "32880:5671")
      .envs("RABBITMQ_DEFAULT_PASS=password1234")
//    .containerName("specific-name")    
//    .leaveRunning(false)
//    .useRunning(false)
//    .labels("com.github.klousiaj.example:a label example")
//    .cleanVolumes(false)
      .waitForLog("Server startup complete")
      .build();

  @Test
  public void testConnectsToDocker() throws Exception {
    ConnectionFactory factory = new ConnectionFactory();
    factory.setHost(rabbitRule.getDockerHost());
    factory.setPort(rabbitRule.getHostPort("5672/tcp"));
    factory.newConnection();
  }
}
```


## Principle

This plugin relies on https://github.com/spotify/docker-client to connect to the docker daemon API.

You can see the latest and greatest build status by checking on the build at travis-ci. Travis CI 
runs the tests against multiple versions of Docker:
 - 1.11.2
 - 1.12.3
 
It has also been validated using docker-toolbox. You should probably set the `DOCKER_HOST` and `DOCKER_CERT_PATH` on your machine.
If they are not set and your are not on UNIX, the client will try to connect to `https://192.168.99.100:2376`,
which is the address of my default docker machine. It works great for me but your mileage may vary.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 1.3.5 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.3.4)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.3.5)
> 2016.12.05

- Implement enhancement #15. Provide the ability for the user to name the container that is being created.
- Provide an example usage that includes a test suite and potential usage during an integration test.
- Upgraded to version 6.1.1 of the docker-client.
- Add ability to add labels to the container that is being created.
- Implement enhancement #16. Provide ability for the user to specify labels at container creation.
- Fix #17. Provide ability for the user to remove volumes automatically at container removal.


## 1.3.4 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.3.4)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.3.4)
> 2016.11.21

- Fix Issue #13. Provide version that doesn't rely on the shaded version of the docker-client.
- Upgraded to version 6.0.0 of the docker-client.
- Removed duplicate dependencies from the build.gradle.

## 1.3.3 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.3.3)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.3.3)
> 2016.11.20

- Fix Issue #11. Added way to use the library without using it as a JUnitRule. 

## 1.3.2 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.3.2)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.3.2)
> 2016.11.10

- Mark WaitForPort as Deprecated. To be removed in v2.0.0.
- Fix Issue #8. Docker for Mac supported without DOCKER_HOST being set.

## 1.3.1 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.3.1)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.3.1)
> 2016-08-04

- Added signing instructions when publishing to Bintray.
- Improved test coverage.

## 1.3.0 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.3.0)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.3.0)
> 2016-08-02

- Add useRunning parameter.
- Add leaveRunning parameter.
- Improve code coverage with mockito.
- No longer validate support for Docker version 1.10.3

## 1.2.2 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.2.2)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.2.2)
> 2016-07-28

- Only pull if the image isn't already available on the Docker Host.

## 1.2.1 [![codecov.io](https://codecov.io/github/klousiaj/docker-junit-rule/coverage.svg?branch=1.2.1)](https://codecov.io/github/klousiaj/docker-junit-rule?branch=1.2.1)
> 2016-07-23

- Add Jacoco and code coverage analysis. (https://www.codecov.io)

## 1.2.0
> 2016-07-21

- Add the ability to specify environment variables to the container.
- Upgrade the docker-client to use version 5.0.2
- Updated travis-ci to test against more versions of Docker.
- Allow for specific ports to be mapped on the HOST during Rule creation.

## 1.1.0
> 2016-07-16

- Add a builder to remove the need for inheritance
- Add a waitForLog directive

## 1.0.2
> 2016-01-06

- Include guava in the dependencies

## 1.0.1
> 2016-01-06

- Use the shaded jar of the docker-client

## 1.0.0
> 2016-01-06

- Initial release


src/test/java/integration/MySqlIntegrationTest.java
package integration;

import com.github.geowarin.junit.DockerRule;
import com.rabbitmq.client.ConnectionFactory;
import org.junit.ClassRule;
import org.junit.Test;

public class MySqlIntegrationTest {

  @ClassRule
  public static DockerRule rabbitRule =
    DockerRule.builder()
      .image("mysql")
      .ports("3306")
      .envs("MYSQL_ROOT_PASSWORD=my_secret")
      .waitForPort("3306/tcp")
//      .waitForLog("")
      .build();

  @Test
  public void testConnectsToDocker() throws Exception {
    // TODO - need to add validation that connects to the DB to ensure that the environment variables are set properly.
  }
}


src/test/java/integration/RabbitIntegrationTest.java
package integration;

import com.github.klousiaj.junit.DockerRule;
import com.rabbitmq.client.Connection;
import com.rabbitmq.client.ConnectionFactory;
import org.junit.Assert;
import org.junit.ClassRule;
import org.junit.Test;

import java.util.Map;

public class RabbitIntegrationTest {

  @ClassRule
  public static DockerRule rabbitRule =
    DockerRule.builder()
      .image("rabbitmq:management")
      .ports("5672", ":32779", "32880:5671")
      .envs("RABBITMQ_DEFAULT_PASS=password1234")
//      .waitForPort("5672/tcp")
      .waitForLog("Server startup complete")
      .build();

  @Test
  public void testConnectsToDocker() throws Exception {
    ConnectionFactory factory = new ConnectionFactory();
    factory.setUsername("guest");
    factory.setPassword("password1234");
    factory.setHost(rabbitRule.getDockerHost());

    Assert.assertNotEquals(-1, rabbitRule.getHostPort("5672/tcp"));
    Assert.assertEquals(32779, rabbitRule.getHostPort("32779/tcp"));
    Assert.assertEquals(32880, rabbitRule.getHostPort("5671/tcp"));
    Assert.assertEquals(-1, rabbitRule.getHostPort("5675/tcp"));

    factory.setPort(rabbitRule.getHostPort("5672/tcp"));
    Connection conn = factory.newConnection();
    Map<String, Object> serverProperties = conn.getServerProperties();
    Assert.assertNotNull(serverProperties);
  }
}


src/test/java/com/github/geowarin/junit/DockerRuleTest.java
package com.github.geowarin.junit;

import com.spotify.docker.client.exceptions.DockerException;
import com.spotify.docker.client.messages.ContainerConfig;
import com.spotify.docker.client.messages.PortBinding;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;

import java.util.List;
import java.util.Map;

import static org.junit.Assert.*;

/**
 * Created by klousiaj on 7/20/16.
 */
public class DockerRuleTest {
  protected final Log logger = LogFactory.getLog(getClass());

  @Test
  public void generatePortBinding() {
    try {
      DockerRule.generatePortBinding("");
      Assert.fail("Should throw an exception");
    } catch (DockerException e) {
      Assert.assertEquals("Invalid port mapping. Was empty or did not match regex /^([\\d]{2,5})?(:\\d{2,5})?$/g. Unable to process ", e.getMessage());
    }

    try {
      DockerRule.generatePortBinding("1");
      Assert.fail("Should throw an exception");
    } catch (DockerException e) {
      Assert.assertEquals("Invalid port mapping. Was empty or did not match regex /^([\\d]{2,5})?(:\\d{2,5})?$/g. Unable to process 1", e.getMessage());
    }

    try {
      DockerRule.generatePortBinding("1:1");
      Assert.fail("Should throw an exception");
    } catch (DockerException e) {
      Assert.assertEquals("Invalid port mapping. Was empty or did not match regex /^([\\d]{2,5})?(:\\d{2,5})?$/g. Unable to process 1:1", e.getMessage());
    }

    try {
      DockerRule.generatePortBinding("1521:");
      Assert.fail("Should throw an exception");
    } catch (DockerException e) {
      Assert.assertEquals("Invalid port mapping. Was empty or did not match regex /^([\\d]{2,5})?(:\\d{2,5})?$/g. Unable to process 1521:", e.getMessage());
    }

    try {
      DockerRule.generatePortBinding("2881721");
      Assert.fail("Should throw an exception");
    } catch (DockerException e) {
      Assert.assertEquals("Invalid port mapping. Was empty or did not match regex /^([\\d]{2,5})?(:\\d{2,5})?$/g. Unable to process 2881721", e.getMessage());
    }

    try {
      DockerRule.generatePortBinding("28/");
      Assert.fail("Should throw an exception");
    } catch (DockerException e) {
      Assert.assertEquals("Invalid port mapping. Was empty or did not match regex /^([\\d]{2,5})?(:\\d{2,5})?$/g. Unable to process 28/", e.getMessage());
    }

    try {
      DockerRule.generatePortBinding("*28");
      Assert.fail("Should throw an exception");
    } catch (DockerException e) {
      Assert.assertEquals("Invalid port mapping. Was empty or did not match regex /^([\\d]{2,5})?(:\\d{2,5})?$/g. Unable to process *28", e.getMessage());
    }

    try {
      Map<String, List<PortBinding>> portBinding = DockerRule.generatePortBinding("4536", "22", "1521");
      Assert.assertEquals(3, portBinding.keySet().size());
      // make sure all of the ports are mapped.
      Assert.assertTrue(portBinding.containsKey("4536"));
      Assert.assertTrue(portBinding.containsKey("22"));
      Assert.assertTrue(portBinding.containsKey("1521"));
      // make sure that they are all generated as random ports
      Assert.assertEquals("", portBinding.get("4536").get(0).hostPort());
      Assert.assertEquals("", portBinding.get("22").get(0).hostPort());
      Assert.assertEquals("", portBinding.get("1521").get(0).hostPort());

    } catch (DockerException e) {
      Assert.fail(e.getMessage());
    }

    try {
      Map<String, List<PortBinding>> portBinding = DockerRule.generatePortBinding(":4536", "9222:22", "1521:1521");
      Assert.assertEquals(3, portBinding.keySet().size());
      // make sure all of the ports are mapped.
      Assert.assertTrue(portBinding.containsKey("4536"));
      Assert.assertTrue(portBinding.containsKey("22"));
      Assert.assertTrue(portBinding.containsKey("1521"));
      // make sure they aren't random, but are actually following the algorithm
      Assert.assertEquals("4536", portBinding.get("4536").get(0).hostPort());
      Assert.assertEquals("9222", portBinding.get("22").get(0).hostPort());
      Assert.assertEquals("1521", portBinding.get("1521").get(0).hostPort());

    } catch (DockerException e) {
      Assert.fail(e.getMessage());
    }

  }
}


.travis.sh
#!/bin/bash
# credit to the guys/gals at spotify who did the real work on this.
# you can find the original here: https://github.com/spotify/docker-client/blob/master/.travis.sh

set -e

if [[ -z $1 ]]; then
  "I need a command!"
  exit 1
fi

case "$1" in
  install_docker)

    if [[ -z $DOCKER_VERSION ]]; then
      echo "DOCKER_VERSION needs to be set as an environment variable"
      exit 1
    fi

    # TODO detect which docker version is already installed and skip
    # uninstall/reinstall if it matches $DOCKER_VERSION

    # stop docker service if running
    sudo stop docker || :

    if [[ "$DOCKER_VERSION" =~ ^1\.12\..* ]]; then
      sudo sh -c 'echo "deb https://apt.dockerproject.org/repo ubuntu-trusty experimental" > /etc/apt/sources.list.d/docker.list'
    fi
    sudo apt-get -qq update
    sudo apt-get -q -y purge docker-engine
    apt-cache policy docker-engine
    sudo apt-get -q -y install docker-engine=$DOCKER_VERSION* "linux-image-extra-$(uname -r)"

    # set DOCKER_OPTS to make sure docker listens on the ports we intend
    echo 'DOCKER_OPTS="-D=true -H=unix:///var/run/docker.sock -H=tcp://127.0.0.1:2375"' | sudo tee -a /etc/default/docker

    # restart the service for the /etc/default/docker change we made after
    # installing the package
    sudo restart docker

    ;;

  dump_docker_config)
    # output the upstart config and default config in case they are needed for
    # troubleshooting
    echo "Contents of /etc/init/docker.conf:"
    sudo cat /etc/init/docker.conf

    echo "Contents of /etc/default/docker"
    sudo cat /etc/default/docker || :

    echo "Contents of /var/log/upstart/docker.log"
    sudo cat /var/log/upstart/docker.log

    ;;

  publish_to_bintray)
    # only upload and publish if it is coming from master
    if [[ $TRAVIS_BRANCH == 'master' ]]; then
      echo "build completed on branch - master. Deploying to bintray"
      ./gradlew bintrayUpload
    fi

    ;;
  *)
    echo "Unknown command $1"
    exit 2
esac


src/main/java/com/github/klousiaj/junit/DockerRuleBuilder.java
package com.github.klousiaj.junit;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;

public class DockerRuleBuilder {
  private final DockerRuleParams params = new DockerRuleParams();

  static final String DEFAULT_LABEL_KEY = "com.github.klousiaj.creator";
  static final String DEFAULT_LABEL_VALUE = "docker-junit-rule";

  private static final String LABEL_PATTERN_STR = "[a-zA-Z0-9\\.-]+:{1}.+";
  static final Pattern LABEL_PATTERN = Pattern.compile(LABEL_PATTERN_STR);


  /**
   * @param imageName The name of the docker image to download
   * @return The builder
   */
  public DockerRuleBuilder image(String imageName) {
    params.imageName = imageName;
    return this;
  }

  /**
   * @param containerName the name specified for the container
   * @return the builder
   */
  public DockerRuleBuilder containerName(String containerName) {
    params.containerName = containerName;
    return this;
  }

  /**
   * <p>
   * List the container ports that you would like to open in the container.
   * There are three ways to specify a port:
   * <ul>
   * <li>443 - this will map the CONTAINER port (443) to a random port on the HOST</li>
   * <li>:443 - this will map the CONTAINER port (443) to the same port on the HOST (443)</li>
   * <li>8443:443 - this will map the CONTAINER port (443) to the specified HOST port (8443)</li>
   * </ul>
   * <p>
   * To know which ports are used on your host use DockerRule#getHostPort(String).
   * Example:
   * <pre>
   * {@code
   * myRule.getHostPort("80/tcp")
   * }
   * </pre>
   *
   * @param ports The ports to open on the container
   * @return The builder
   * @see DockerRule#getHostPort(String)
   */
  public DockerRuleBuilder ports(String... ports) {
    params.ports = ports;
    return this;
  }

  /**
   * <p>
   * The environment variables to be passed to the container.
   * </p>
   * <pre>
   *     {@code
   *      MYSQL_ROOT_PASSWORD=my_secret
   *     }
   * </pre>
   *
   * @param envs The environments variables to pass to the container.
   * @return The builder
   */
  public DockerRuleBuilder envs(String... envs) {
    params.envs = envs;
    return this;
  }

  /**
   * Leave the container running rather than shutting down and removing the
   * container.
   * <p>
   * <b>Note</b> - use this feature with caution. If misused, test cases may
   * have inconsistent results depending on the value of this parameter. Depending on
   * your use case, a better solution may be to use a JUnit Test @Suite.
   *
   * @param leaveRunning true if the container should be left running after test execution
   * @return the builder
   */
  public DockerRuleBuilder leaveRunning(boolean leaveRunning) {
    params.leaveRunning = leaveRunning;
    return this;
  }

  /**
   * If a container of the requested image/version is already running on the requested
   * port an attempt will be made to use that container.
   * <p>
   * <b>Note</b> - use this feature with caution. If misused, test cases may
   * have inconsistent results depending on the value of this parameter. Depending on
   * your use case, a better solution may be to use a JUnit Test @Suite.
   *
   * @param useRunning true if a running container should be used if found
   * @return the builder
   */
  public DockerRuleBuilder useRunning(boolean useRunning) {
    params.useRunning = useRunning;
    return this;
  }

  /**
   * Allow the user to specify whether a volume associated with the requested image
   * should be removed as part of the cleanup.
   * <p>
   * <b>Note</b> - this is defaulted to false to ensure backward compatibility. This is likely
   * not the desired default behavior. In the future this functionality may be changed to default
   * to true.
   *
   * @param cleanVolumes true if volumes should be removed
   * @return the builder
   */
  public DockerRuleBuilder cleanVolumes(boolean cleanVolumes) {
    params.cleanVolumes = cleanVolumes;
    return this;
  }

  /**
   * A convenience method that allows for labels to be specified as concatenated strings in the form:
   * <code>key:value</code>.
   * <p>
   * Must match the PATTERN: [a-zA-Z0-9\\.-]+:{1}.+
   *
   * @param labels the key:value pairs to be applied to the created container.
   * @return the builder
   * @throws IllegalArgumentException if the strings cannot be parsed to match a key-value pair
   */
  public DockerRuleBuilder labels(String... labels) throws IllegalArgumentException {
    if (labels != null) {
      Map<String, String> labelMap = new HashMap<>();
      for (String label : labels) {
        if (!LABEL_PATTERN.matcher(label).matches()) {
          throw new IllegalArgumentException();
        }
        String[] split = label.split(":");
        labelMap.put(split[0], split[1]);
      }
      if (labelMap.size() > 0) {
        return labels(labelMap);
      }
    }
    return this;
  }

  /**
   * Specify labels to associate with this container. By default, the label com.github.klousiaj.creator:docker-junit-rule
   * will be added. Attempts to override this will be ignored.
   *
   * @param labels the labels to add
   * @return the builder
   */
  protected DockerRuleBuilder labels(Map<String, String> labels) {
    params.labels = labels;
    return this;
  }

  public DockerRuleBuilder cmd(String cmd) {
    params.cmd = cmd;
    return this;
  }

  /**
   * Utility method to ensure a container is started
   *
   * @param portToWaitOn The port to wait on
   * @return The builder
   */
  public DockerRuleBuilder waitForPort(String portToWaitOn) {
    return waitForPort(portToWaitOn, 10000);
  }

  /**
   * Utility method to ensure a container is started
   *
   * @param portToWaitOn    The port to wait on
   * @param timeoutInMillis Maximum waiting time in milliseconds
   * @return The builder
   */
  @Deprecated
  public DockerRuleBuilder waitForPort(String portToWaitOn, int timeoutInMillis) {
    params.portToWaitOn = portToWaitOn;
    params.waitTimeout = timeoutInMillis;
    return this;
  }

  /**
   * @param logToWait The log message to wait.
   *                  This will stop blocking as soon as the docker logs contains that string
   * @return The builder
   */
  public DockerRuleBuilder waitForLog(String logToWait) {
    params.logToWait = logToWait;
    return this;
  }

  public DockerRule build() {
    return new DockerRule(params).initialize();
  }

}


src/main/java/com/github/klousiaj/junit/DockerRule.java
package com.github.klousiaj.junit;

import com.spotify.docker.client.*;
import com.spotify.docker.client.exceptions.DockerCertificateException;
import com.spotify.docker.client.exceptions.DockerException;
import com.spotify.docker.client.exceptions.ImageNotFoundException;
import com.spotify.docker.client.messages.*;
import org.apache.commons.lang.StringUtils;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.junit.rules.ExternalResource;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.InetSocketAddress;
import java.net.SocketAddress;
import java.net.URI;
import java.nio.ByteBuffer;
import java.nio.channels.SocketChannel;
import java.nio.file.Paths;
import java.util.*;
import java.util.regex.Pattern;

import static com.spotify.docker.client.DockerClient.LogsParam.follow;
import static com.spotify.docker.client.DockerClient.LogsParam.stdout;

/**
 * <p>
 * JUnit rule starting a docker container before the test and killing it
 * afterwards.
 * </p>
 * <p>
 * Uses spotify/docker-client.
 * Adapted from https://gist.github.com/mosheeshel/c427b43c36b256731a0b
 * </p>
 * author: Geoffroy Warin (geowarin.github.io)
 */
public class DockerRule extends ExternalResource {
  private static final Log logger = LogFactory.getLog(DockerRule.class);
  private static final String DOCKER_MACHINE_SERVICE_URL = "https://192.168.99.100:2376";
  private static final String PORT_MAPPING_REGEX = "^([\\d]{2,5})?(:\\d{2,5})?$";
  static final String CONTAINER_NAME_REGEX = "/?[a-zA-Z0-9_-]+";
  private static final Pattern PORT_PATTERN = Pattern.compile(DockerRule.PORT_MAPPING_REGEX);
  private static final Pattern NAME_PATTERN = Pattern.compile(DockerRule.CONTAINER_NAME_REGEX);

  DockerRuleParams params;

  private final DockerClient dockerClient;
  private ContainerCreation container;
  private Map<String, List<PortBinding>> ports;

  public static DockerRuleBuilder builder() {
    return new DockerRuleBuilder();
  }

  DockerRule(DockerClient client) {
    this.dockerClient = client;
  }

  DockerRule(DockerRuleParams params) {
    this.params = params;
    dockerClient = createDockerClient();
  }

  public DockerRule initialize() {
    try {
      try {
        // try to use a local copy of the image if one exists
        ImageInfo imageInfo = dockerClient.inspectImage(params.imageName);
      } catch (ImageNotFoundException e) {
        logger.info("Unable to find the requested image locally. Will attempt to pull from docker hub.");
        dockerClient.pull(params.imageName);
      }
    } catch (DockerException | InterruptedException e) {
      throw new IllegalStateException(e);
    }
    return this;
  }

  @Override
  protected void before() throws Throwable {
    super.before();
    this.start();
  }

  /**
   * Start method.
   * Allows for the functionality to be used without the overhead of the jUnit Rule.
   *
   * @throws Throwable throwable exception
   */
  public void start() throws Throwable {
    // attach to a running container, or create and start one
    attachToContainer();

    if (params.portToWaitOn != null) {
      waitForPort(getHostPort(params.portToWaitOn), params.waitTimeout);
    }

    if (params.logToWait != null) {
      waitForLog(params.logToWait);
    }
  }

  @Override
  protected void after() {
    super.after();
    this.stop();
  }

  /**
   * Stop method.
   * Allows for the functionality to be used without the overhead of the JUnit rule.
   */
  public void stop() {
    if (!params.leaveRunning) {
      try {
        try {
          dockerClient.killContainer(getContainer().id());
        } catch (DockerException | InterruptedException e) {
          logger.error("Unable to stop docker container " + getContainer().id(), e);
          logger.info("Will attempt to remove container anyway.");
        }
        dockerClient.removeContainer(getContainer().id(),
          DockerClient.RemoveContainerParam.removeVolumes(params.cleanVolumes));
      } catch (DockerException | InterruptedException e) {
        throw new RuntimeException("Unable to remove docker container " + getContainer().id(), e);
      } finally {
        dockerClient.close();
      }
    }
  }

  /**
   * Utility method to get the docker host.
   * Can be different from localhost if using docker-machine
   *
   * @return The current docker host
   */
  public String getDockerHost() {
    return dockerClient.getHost();
  }

  /**
   * Get the name of the name of the Docker container being used by this Rule
   *
   * @return the name of the Docker container
   * @throws DockerException      a DockerException
   * @throws InterruptedException an InterruptedException
   */
  public String getContainerName() throws DockerException, InterruptedException {
    return dockerClient.inspectContainer(container.id()).name().substring(1);
  }

  /**
   * For the provided CONTAINER port, return the mapped HOST port.
   *
   * @param containerPort the containerPort - i.e. 80/TCP is
   * @return -1 if the port is not mapped
   */
  public int getHostPort(String containerPort) {
    List<PortBinding> portBindings = ports.get(containerPort);
    if (portBindings == null || portBindings.isEmpty()) {
      return -1;
    }
    return Integer.parseInt(portBindings.get(0).hostPort());
  }

  /**
   * set the container for the test run by either attaching to an already running
   * container or by creating and starting a brand new one.
   *
   * @throws DockerException
   * @throws InterruptedException
   */
  void attachToContainer() throws DockerException, InterruptedException {
    Map<String, List<PortBinding>> portBinding = generatePortBinding(params.ports);
    String containerId = null;
    if (params.useRunning) {
      containerId = foundRunningContainer(portBinding);
    }
    if (containerId == null) {
      // configure the container based on the provided parameters
      ContainerConfig containerConfig = createContainerConfig(params.imageName,
        portBinding, params.envs, params.cmd, params.labels);

      if (isValidContainerName(params.containerName))
        container = dockerClient.createContainer(containerConfig, params.containerName);
      else
        container = dockerClient.createContainer(containerConfig);

      dockerClient.startContainer(getContainer().id());
    } else {
      logger.warn("Connecting to an already running container (" + containerId + "). Please note this is not the default behavior and should only be used by advanced users.");
      this.container = new ContainerCreation(containerId);
    }

    ContainerInfo info = dockerClient.inspectContainer(getContainer().id());
    ports = info.networkSettings().ports();
  }

  /**
   * Ensure that the provided containerName is one that will be accepted by
   * the Docker daemon. An empty or null string is considered invalid. Additionally
   * a name that does not match the value of DockerRule.CONTAINER_NAME_REGEX is
   * considered invalid.
   *
   * @param containerName the name to be validated
   * @return true if provided container name can be used
   */
  boolean isValidContainerName(String containerName) {
    if (StringUtils.isEmpty(containerName)) {
      logger.trace("no user provided container name was provided");
      return false;
    }
    if (!NAME_PATTERN.matcher(containerName).matches()) {
      logger.error("provided container name " + containerName + " doesn't match regex " + DockerRule.CONTAINER_NAME_REGEX);
      logger.info("container name will be generated");
      return false;
    }
    return true;
  }

  /**
   * Search for an already running container and set the container object if one is found
   *
   * @return true if a container with the same image:tag/port is already running
   */
  String foundRunningContainer(Map<String, List<PortBinding>> bindings) throws DockerException, InterruptedException {
    for (Container instanceContainers : dockerClient.listContainers()) {
      // only consider running containers of the same image.
      if (params.imageName.equals(instanceContainers.image()) && "running".equals(instanceContainers.state())) {
        // the requested ports count must be less than or equal to the existing containers exposed port count
        if (instanceContainers.ports().size() >= bindings.size()) {
          boolean portsAlign = true;
          Iterator<String> iter = bindings.keySet().iterator();
          while (iter.hasNext() && portsAlign) {
            String key = iter.next();
            portsAlign = this.validPortMap(key, bindings.get(key).get(0).hostPort(), instanceContainers.ports());
          }
          if (portsAlign) {
            return instanceContainers.id();
          }
        }
      }
    }
    return null;
  }

  boolean validPortMap(String privatePort, String publicPort, List<Container.PortMapping> containerMapping) {
    boolean mappingFound = false;
    for (int ii = 0; ii < containerMapping.size() && !mappingFound; ii++) {
      String containerPort = String.valueOf(containerMapping.get(ii).getPrivatePort());
      String hostPort = String.valueOf(containerMapping.get(ii).getPublicPort());
      if (containerPort.equals(privatePort) && ("".equals(publicPort) || publicPort.equals(hostPort))) {
        mappingFound = true;
      }
    }
    return mappingFound;
  }

  /**
   * Wait for the requested port to be accessible.
   *
   * @param port            the port number
   * @param timeoutInMillis the time to wait before failing
   */
  @Deprecated
  public void waitForPort(int port, long timeoutInMillis) {
    SocketAddress address = new InetSocketAddress(getDockerHost(), port);
    long totalWait = 0;
    while (true) {
      try {
        SocketChannel.open(address);
        return;
      } catch (IOException e) {
        try {
          Thread.sleep(100);
          totalWait += 100;
          if (totalWait > timeoutInMillis) {
            throw new IllegalStateException("Timeout while waiting for port " + port);
          }
        } catch (InterruptedException ie) {
          throw new IllegalStateException(ie);
        }
      }
    }
  }

  protected DockerClient createDockerClient() {
    if (isUnix() || System.getenv("DOCKER_HOST") != null) {
      try {
        return DefaultDockerClient.fromEnv().build();
      } catch (DockerCertificateException e) {
        System.err.println(e.getMessage());
      }
    }

    logger.info("Could not create docker client from the environment. Assuming docker-machine environment with url " + DOCKER_MACHINE_SERVICE_URL);
    DockerCertificates dockerCertificates = null;
    try {
      String userHome = System.getProperty("user.home");
      dockerCertificates = new DockerCertificates(Paths.get(userHome, ".docker/machine/certs"));
    } catch (DockerCertificateException e) {
      System.err.println(e.getMessage());
    }
    return DefaultDockerClient.builder()
      .uri(URI.create(DOCKER_MACHINE_SERVICE_URL))
      .dockerCertificates(dockerCertificates)
      .build();
  }

  protected ContainerConfig createContainerConfig(String imageName, Map<String, List<PortBinding>> portBinding,
                                                  String[] envs, String cmd, Map<String, String> labels) throws DockerException {
    HostConfig hostConfig = HostConfig.builder()
      .portBindings(portBinding)
      .build();

    // make sure the labels includes the default label information
    if (labels != null) {
      labels.put(DockerRuleBuilder.DEFAULT_LABEL_KEY, DockerRuleBuilder.DEFAULT_LABEL_VALUE);
    }

    ContainerConfig.Builder configBuilder = ContainerConfig.builder()
      .hostConfig(hostConfig)
      .image(imageName)
      .env(envs)
      .networkDisabled(false)
      .labels(labels)
      .exposedPorts(portBinding.keySet());

    if (cmd != null) {
      configBuilder = configBuilder.cmd(cmd);
    }
    return configBuilder.build();
  }

  protected static Map<String, List<PortBinding>> generatePortBinding(String... ports) throws DockerException {

    Map<String, List<PortBinding>> portBindings = new HashMap<>();
    for (String port : ports) {
      // check to see if the port is able to be parsed
      if ("".equals(port) || !PORT_PATTERN.matcher(port).matches())
        throw new DockerException("Invalid port mapping. Was empty or did not match regex /" + PORT_MAPPING_REGEX + "/g. Unable to process " + port);

      // it will be random port if there is not a colon included
      PortBinding binding = PortBinding.randomPort("0.0.0.0");
      String[] split = port.split(":");
      if (split.length > 1) {
        // clean up the container port description
        port = split[1];
        binding = "".equals(split[0]) ? PortBinding.of("0.0.0.0", split[1]) : PortBinding.of("0.0.0.0", split[0]);
      }
      List<PortBinding> hostPorts = Collections.singletonList(binding);
      portBindings.put(port, hostPorts);
    }
    return portBindings;
  }

  private static boolean isUnix() {
    String os = System.getProperty("os.name").toLowerCase();
    return os.contains("nix") || os.contains("nux") || os.contains("aix") || os.equals("mac os x");
  }

  protected void waitForLog(String messageToMatch) throws DockerException, InterruptedException, UnsupportedEncodingException {
    LogStream logs = dockerClient.logs(getContainer().id(), follow(), stdout());
    String log;
    do {
      LogMessage logMessage = logs.next();
      ByteBuffer buffer = logMessage.content();
      byte[] bytes = new byte[buffer.remaining()];
      buffer.get(bytes);
      log = new String(bytes);
    } while (!log.contains(messageToMatch));
  }

  protected ContainerCreation getContainer() {
    return this.container;
  }

  protected List<Container> getContainerByLabel(String labelKey) throws DockerException, InterruptedException {
    return this.getContainerByLabel(labelKey, null);
  }

  protected List<Container> getContainerByLabel(String labelKey, String labelValue) throws DockerException, InterruptedException {
    return dockerClient.listContainers(DockerClient.ListContainersParam.withLabel(labelKey, labelValue));
  }
}
